### ISSUE TITLE
Linux perf does not consume the current debuginfo format

### ISSUE THREAD

**Issue by adinn** [oracle/graal#4810]
**Describe the issue**
On Linux the perf tool provides details information about code execution and performance. It should be able to use image debuginfo to derive function names, files and line numbers for compiled method code, including inlined method code. There are several features of the current GraalVM native debug info layout that, while valid as DWARF and recognized by gdb, do not match perf's expectations. The debuginfo layout should be modified to accommodate perf's expectations as follows: 

1. Method declarations need to have an associated line number in order for perf to recognize them (we can use whatever line is associated with bytecode zero)
2. `.debug_arange` section entries need to be sorted low to high by range start address (this was always meant to be the case ...)
3. Concrete inline method definitions need to refer to an abstract inline method declaration replicated inliner's compile unit (CU) for the inliner's class using a CU local offset. At present they refer to a single definition in the CU of the class that declares the inlined method using a section global offset.
4. Inline method declarations also need to replicate their original method's parameters/local declarations in the importing class's CU so that the latter can be referred to from local/parameter location definitions using a CU local offset. At present local/parameter location definitions refer to a single, corresponding local/parameter declaration in the CU of the class that declares the inlined method.
5. CUs for instances classes whose methods only exist inline need to include a `stmt_list` attribute identifying an associated `.debug_line` subsection detailing line info for the class's methods. Clearly, this subsection will include no actual line number records. However, perf requires it so as to provide a dir and file table identifying the source path and file for the CU's class.
6. Perf expects method declarations to include a `linkage_name` attribute referencing a string section entry populated with the fully qualified method name (formatted using format specifier `"%H.%n"`). Unlike gdb, perf will not synthesize method names from other DWARF data. If a `linkage_name` attribute is omitted, perf picks up the symbol table entry with the nearest address below a profile target as a label for the corresponding function.
n.b. This field is also used by the debugger to present the method in debug backtraces etc, meaning that this patch results in some minor cosmetic changes to debug output. This also means that it would a very bad idea to use the GraalVM synthesized symbol name as the value of the linkage_name attribute (includes a message digest encoding of the arguments). 

**Steps to reproduce the issue**
Use `perf record -g` to record profile events for a generated native built with debug info enabled.
Use `perf report` to view the resulting profile info.
Java code events are reported either using a hex address or as if they belong to an unrelated C++ function such as _start.

**Describe GraalVM and your environment:**
Latest GraalVM head on Linux.

**Comment by adinn** [comment]
I found another constraint that needs to be met for proper perf reporting.

7. perf will not translate a sample code address that lies in the `.text` section to a DWARF-derived method/function name if it precedes the first symbol defined in the .text section. The region preceding that symbol is assumed to be dead code.
Ditto for the other code sections (`.plt`/`.init`).

So, if you build your image with (default) setting `-H:+DeleteLocalSymbols` then samples that precede the exported C entry point for a method of class `VM_Locator` get reported as hex addresses.

If you build with `-H:-DeleteLocalSymbols` then every method start is marked by a symbol so all samples get reported using a DWARF-derived method name.

I have pushed an update which responds to this problem. If `DeleteLocalSymbols` is true it inserts a dummy function symbol at the start of the first compiled method in the .text section, ensuring all methods have their names derived.
